step5__run_haplotype_caller_per_sample.py

```diff
@@ -1,88 +1,142 @@
-				continue
-
-			counter += 1
-			if args.num_samples_to_process and counter > args.num_samples_to_process:
-				break
-
-			j = batch_utils.init_job(batch, f"readviz: {row.sample_id}", DOCKER_IMAGE if not args.raw else None, args.cpu, args.memory)
-			batch_utils.switch_gcloud_auth_to_user_account(j, GCLOUD_CREDENTIALS_LOCATION, GCLOUD_USER_ACCOUNT)
-
-			local_exclude_intervals = batch_utils.localize_file(j, EXCLUDE_INTERVALS)
-			local_fasta = batch_utils.localize_file(j, batch_utils.HG38_REF_PATHS.fasta, use_gcsfuse=True)
-			local_fasta_fai = batch_utils.localize_file(j, batch_utils.HG38_REF_PATHS.fai, use_gcsfuse=True)
-			batch_utils.localize_file(j, batch_utils.HG38_REF_PATHS.dict, use_gcsfuse=True)
-			local_tsv_bgz = batch_utils.localize_file(j, row.variants_tsv_bgz)
-			local_cram_path = batch_utils.localize_file(j, row.cram_path)
-			local_crai_path = batch_utils.localize_file(j, row.crai_path)
-
-			j.command(f"""echo --------------
+    """
+    Run HaplotypeCaller to generate bamouts.
+
+    Step 5 of readviz pipeline.
+    """
+    _, args = parse_args()
+    hl.init(log="/dev/null", quiet=True)
+    project = args.gcloud_project
+    docker_image = args.docker_image
+    set_gcloud_project(project)
+
+    logger.info("Making sure input cram_and_tsv_paths_table arg is valid...")
+    bams = {}
+    samples = {}
+    with hl.hadoop_open(args.cram_and_tsv_paths_table) as c:
+        # Confirm header has all required columns
+        header = c.readline().strip().split("\t")
+        if {"sample_id", "cram_path", "crai_path", "variants_tsv_bgz"} - set(header):
+            raise DataException(
+                "%s must contain 'sample_id', 'cram_path', 'crai_path', variants_tsv_bgz' columns!"
+            )
+
+        for line in c:
+            sample, cram, crai, variants_tsv_bgz = line.strip().split("\t")
+
+            # Store output BAM path
+            bam = f"{args.output_prefix}/{sample}.bamout.bam"
+            bai = f"{args.output_prefix}/{sample}.bamout.bai"
+            bams[sample] = bam
+
+            # Store sample information
+            samples[sample] = [cram, crai, variants_tsv_bgz, bam, bai]
+
+            logger.info(
+                "Checking that all input crams are 'US-CENTRAL1' or multi-regional buckets..."
+            )
+            # Check that all buckets are in "US-CENTRAL1" or are multi-regional to avoid egress charges to the Batch cluster
+            check_storage_bucket_region(cram)
+
+    logger.info("Checking if any output bams already exist...")
+    bam_files_exist = call_parallel_file_exists(list(bams.values()))
+
+    samples_without_bams = []
+    for sample in bams:
+        if not bam_files_exist[bams[sample]]:
+            samples_without_bams.append(sample)
+
+    # Process samples
+    with run_batch(args, batch_name=f"HaplotypeCaller -bamout") as batch:
+        for _ in tqdm(range(len(samples)), desc="submit HC batch job for samples",):
+            for sample in samples:
```

Comment:
wouldn't you want to only process samples_without_bams here?

Reply:
oops yes!